### ISSUE TITLE
Architecture: a fair, repeatable eval-and-improve loop for SKILL.md files driven by `claude -p`

### ISSUE THREAD

**Issue by rosenhouse** [rosenhouse/skill-refinery#1]
## The question this issue answers

> I'm trying to get Claude to help me fairly evaluate a skill definition markdown file, and improve it, in a loop. Maybe we build a small program that drives `claude -p` locally to run a given skill against a set of experimental inputs with controlled context, runs `claude -p` again with an evaluation prompt (LLM as judge), and then a separate step edits the skill definition. Then a meta-skill operates that process in a loop with deterministic constraints and context management. Does this make sense? Are there existing open-source projects that do this well? If not, what architecture?

## Summary

Yes, the idea makes sense, and it is close to how Anthropic's own tooling already works. The three-step loop (run the skill against fixed inputs, judge the outputs with a separate LLM call, edit the skill) is sound. The hard part is not the loop itself but making the *measurement* trustworthy enough that an automated editor cannot game it. This issue records the survey of what exists, the recommended architecture, and a build plan.

The short version:

- **Build on Anthropic's primitives, not a generic prompt-optimization framework.** Claude Code already ships a run-plus-judge harness (`claude plugin eval`, early access), a hermetic headless mode (`claude -p --bare`), structured output (`--json-schema`), and a reference implementation of the improve loop (the `skill-creator` skill in `anthropics/skills`). Nothing in the DSPy / promptfoo / TextGrad family understands a SKILL.md, a `Skill` tool call, or an agentic transcript with file outputs.
- **The gap is in fairness and consistency, not plumbing.** `skill-creator` automates the loop only for the frontmatter *description* (trigger accuracy). Improving the skill *body* is human-in-the-loop, with no held-out set, no judge calibration, no blinding, and no regression gate. That is exactly the piece to build here.
- **Recommended shape:** a small deterministic runner (Python, no framework) with three isolated stages and a frozen judge, plus a meta-skill that drives it but is structurally prevented from seeing the held-out set or the judge's private rubric.
- **Two constraints to know before building.** `claude -p` has no temperature or seed, so repeatability has to come from hermetic context and repeated sampling. And `--bare`, the flag that gives hermetic context, authenticates only with an API key, never OAuth, so subscription-only users need the fallback flags described under Stage 1.

## What exists today

Surveyed as of 2026-09-10. Verified against the local CLI (`claude` 2.1.267) where possible; items marked *unverified* are from docs or READMEs only.

### Anthropic tooling (the parts to build on)

| Thing | What it covers | What it does not |
|---|---|---|
| [`skill-creator`](https://github.com/anthropics/skills/tree/main/skills/skill-creator) (Apache-2.0) | The whole loop, described in prose. Runs with-skill and baseline arms as subagents, an LLM grader against per-case assertions (`agents/grader.md`), a blind pairwise comparator (`agents/comparator.md`), an un-blinding analyzer, and `aggregate_benchmark.py` for mean and stddev. `run_loop.py` + `run_eval.py` + `improve_description.py` are a real, scripted, `claude -p`-driven eval-improve loop with a stratified 60/40 train and held-out split, 3 runs per query, and best-by-held-out selection. | The scripted loop only optimizes the frontmatter `description` (trigger accuracy). Body improvement is human-in-the-loop via subagents, with no hermetic isolation, no held-out set, no judge calibration, and no accept/reject gate. |
| `claude plugin eval` (early access, gated per org; help text visible locally, `init` refuses to run here) | Native run-plus-judge harness. Cases under `evals/<case>/prompt.md` plus `graders/*.md`. Graders: `regex`, `tool_used`, `tool_order`, `file_exists`, `llm` (2-of-3 voting), `baseline`. Flags: `--runs` (default 3), `--ablation with-without` for a no-plugin baseline arm with score delta, `--judge-model`, `--threshold`, `--max-cost-usd`, `--mocks` for MCP servers, `--json`, HTML `--report`. | No improve step. Not documented on code.claude.com yet, so the case format could change. |
| `claude -p` headless flags | `--bare`, `--add-dir`, `--plugin-dir`, `--tools`, `--allowedTools`, `--permission-mode`, `--permission-prompts none`, `--max-turns`, `--max-budget-usd`, `--json-schema`, `--output-format json\|stream-json`, `--no-session-persistence`, `--model`, `--effort`. The `json` result carries `total_cost_usd`, `num_turns`, `usage`, `session_id`, `structured_output`. `stream-json` emits a `system/init` event listing loaded tools, plugins, and MCP servers. | No temperature or seed ([anthropics/claude-code#3370](https://github.com/anthropics/claude-code/issues/3370)). `--bare` requires `ANTHROPIC_API_KEY` or an `apiKeyHelper`; it never reads OAuth credentials, so subscription-only users need the fallback in the run stage below. |
| `/skill-doctor` | Context-cost and invocation-frequency report per skill. | Not a quality eval. |
| Claude Agent SDK | Same option surface as `claude -p`, in-process. | No eval or optimize layer. Not needed unless the runner wants in-process hooks. |

### Generic prompt-optimization frameworks

None of these know what a SKILL.md is, and most cannot treat an arbitrary subprocess as the system under test.

| Tool | CLI as system under test | Improve step | Verdict |
|---|---|---|---|
| [DSPy](https://dspy.ai/api/optimizers/GEPA/overview/) MIPROv2 / GEPA | No. Optimizes `dspy.Module` predictors. | Yes | Wrong shape for us. |
| [gepa-ai/gepa](https://github.com/gepa-ai/gepa) (MIT) | Yes via a custom `GEPAAdapter`; nothing stops the evaluator shelling out to `claude -p`. | Yes, reflect-mutate-Pareto. | The most reusable *engine* if we later want a smarter proposer than "ask Claude for a new draft." Not needed for v1. |
| [promptfoo](https://www.promptfoo.dev/docs/usage/prompt-optimization/) (MIT) | Yes, `exec:` provider. | `promptfoo optimize` edits a prompt string, with a validation split. *Unverified* whether optimize works with `exec:` providers. | Good judge-assertion library; optimize step targets a string, not a file with bundled resources. |
| [DeepEval](https://deepeval.com/docs/prompt-optimization-introduction) `PromptOptimizer` | Yes, arbitrary Python callback. | Yes (GEPA default). | Same limitation as promptfoo: optimizes a string. |
| [Promptim](https://github.com/hinthornw/promptimizer), [TextGrad](https://github.com/zou-group/textgrad) | Partial. | Yes | LangSmith-centric / research code. |
| [Inspect AI](https://inspect.aisi.org.uk/agent-bridge.html) | Yes. `sandbox_agent_bridge()` runs a CLI agent in a sandbox and proxies its API calls. | No | Best-in-class *measurement* harness for Claude Code if we ever need sandboxing and full API-call capture. Overkill for v1. |
| OpenAI evals, Braintrust autoevals, Ragas, Langfuse, Weave | Yes | No | Scorer libraries and dashboards. |

### Agent-Skills-specific projects

- [UiPath/coder_eval](https://github.com/uipath/coder_eval) (Apache-2.0, ~130 stars): YAML suites, sandboxed Claude Code / Codex / Gemini runs, `skill_triggered` checks, LLM-rubric plus deterministic graders, A/B experiments, CI gates. Measurement only, but the closest thing to a mature stage-1-plus-stage-2 outside Anthropic.
- [TiesPetersen/SkillBenchmark](https://github.com/TiesPetersen/SkillBenchmark) (MIT, ~14 stars): with/without via the raw API (skill injected as a system prompt, not Claude Code), blind judge, Welch interval on the delta. Measurement only.
- [rartzi/ClaudeSkills-Optimizer-GEFA](https://github.com/rartzi/ClaudeSkills-Optimizer-GEFA) (MIT, ~3 stars): GEPA-driven SKILL.md evolution using Claude Code runs. The only public project doing all three steps. Very immature, no held-out split documented.
- [SkillOpt](https://arxiv.org/abs/2605.23904) (research, code at aka.ms/skillopt, not inspected): bounded add/delete/replace edits to a skill doc, accepted only if validation score improves, with a rejected-edit buffer. This is the closest research precedent for the gate described below.
- [agentskills.io evaluation guide](https://agentskills.io/skill-creation/evaluating-skills): fresh session per run, with/without baseline, assertions written after seeing first outputs, blind comparison for version-vs-version, drop non-discriminating assertions, then hand signals plus the current SKILL.md to an LLM to propose changes. Matches the proposed design; describes but does not script it.

### Bottom line of the survey

Nothing mature does all three steps against Claude Code as the runner. Measurement is well covered (`skill-creator`, `plugin eval`, coder_eval, Inspect). The improve step exists generically (gepa, DeepEval, promptfoo optimize) and for skills only in a 3-star repo and a paper. `skill-creator`'s `run_loop.py` is the nearest sanctioned template for a `claude -p` loop and already embodies train/held-out selection. Build a small runner that borrows its conventions, and keep the case format close enough to `plugin eval`'s that a later swap is cheap.

### Judge methodology worth encoding, with sources

- **Pairwise beats absolute for accept/reject; absolute is better for trend lines.** Pairwise has higher human agreement (MT-Bench reports ~85%) but suffers position bias. Swap A/B order and require agreement ([Judging the Judges, arXiv 2406.07791](https://arxiv.org/html/2406.07791v5)). Rubric option *order* also biases outcomes ([arXiv 2602.02219](https://arxiv.org/abs/2602.02219)).
- **Self-preference and family bias are real.** Judges favor their own and same-family outputs ([arXiv 2410.21819](https://arxiv.org/abs/2410.21819), [arXiv 2604.22891](https://arxiv.org/html/2604.22891v4)). Both our runner and judge are Claude, so at minimum use a different Claude model for judging, hide which arm is new, and spot-check a subset with a non-Anthropic judge.
- **Calibrate against humans once per experiment.** Hand-label a small set, compute agreement (Cohen's kappa), revise the rubric until disagreements are explainable ([Ragas align-judge guide](https://docs.ragas.io/en/stable/howtos/applications/align-llm-as-judge/)).
- **Every serious optimizer accepts an edit only on a held-out split** (GEPA, promptfoo optimize, Promptim, `skill-creator` `run_loop.py`, SkillOpt). Drop assertions that pass in both arms.

## Why "fair and consistent" is the actual problem

An automated improve loop has three failure modes that a naive implementation hits immediately:

1. **Goodhart on the judge.** The editor learns what the judge rewards (length, confident tone, restating the rubric) rather than what the user wants. Fix: the editor never sees the judge prompt, the judge grades against pre-written per-case assertions rather than an open "is this good" question, and a held-out set is scored but never shown to the editor.
2. **Overfitting to the training cases.** After five iterations the SKILL.md contains a bulleted list of the five test prompts. Fix: train/held-out split from the start, accept a candidate only if held-out score does not regress, and an explicit "no case-specific instructions" rule in the editor prompt (the `skill-creator` improve prompt already says this well).
3. **Noise mistaken for signal.** A single run of a stochastic agent per case cannot distinguish a 10-point improvement from a coin flip. Fix: N runs per case per candidate (three minimum), report mean and spread, and only accept a candidate when the improvement clears a pre-set margin on both train and held-out.

There is no temperature or seed control in `claude -p`, so determinism has to come from everything *around* the model: hermetic context, pinned model, pinned skill and judge versions, fixed inputs, and repeated sampling.

## Recommended architecture

Three stages, each a separate `claude -p` process, each with its own tool allowlist, model pin, and budget. A thin Python runner sequences them and owns all state on disk. The meta-skill talks only to the runner.

```
skill-refinery/
  refinery/                 # Python package, the runner (no LLM calls of its own)
    run.py                  # stage 1: execute one case against one skill version
    judge.py                # stage 2: grade one run against its assertions
    propose.py              # stage 3: ask the editor for a new SKILL.md
    loop.py                 # orchestrates: run -> judge -> propose -> gate -> repeat
    report.py               # markdown/HTML summary per iteration
  skills/<name>/            # the skill under test (git-tracked; each candidate is a commit)
  evals/<name>/
    cases/<case-id>/
      prompt.md             # the user prompt for this case
      inputs/               # fixture files copied into a fresh workdir
      assertions.yaml       # objective, per-case checks the judge grades
    judge.md                # the judge prompt, frozen per experiment
    manifest.yaml           # model pins, runs-per-case, thresholds, budgets,
                            # and the train|holdout split (never shown to the editor)
  runs/<experiment>/<iter>/<case>/<run-n>/
    workdir/                # fresh copy of inputs, all outputs land here
    transcript.jsonl        # stream-json from claude -p
    result.json             # the final json result (cost, turns, session id)
    grade.json              # judge output, structured via --json-schema
```

### Stage 1: run

One `claude -p` per (case, candidate, run index). Hermetic by construction:

```
claude -p --bare \
  --model <pinned> \
  --add-dir <workdir> \
  --plugin-dir <candidate-skill-as-plugin> \
  --tools "Read,Write,Edit,Bash,Glob,Grep" \
  --permission-mode acceptEdits --permission-prompts none \
  --max-turns 40 --max-budget-usd 1.00 \
  --no-session-persistence \
  --output-format stream-json --verbose \
  "$(cat prompt.md)"
```

Key points:

- `--bare` skips hooks, auto-memory, CLAUDE.md discovery, plugin sync, and keychain reads. Whatever context the skill gets is passed explicitly. This is the single most important flag for reproducibility. Caveat: `--bare` authenticates only via `ANTHROPIC_API_KEY` or an `apiKeyHelper`, never OAuth. For subscription-only users the fallback is `--setting-sources "" --strict-mcp-config --disable-slash-commands` plus an explicit `--plugin-dir`, and the runner should assert hermeticity either way by parsing the `system/init` event in `stream-json` and failing the run if any unexpected skill, plugin, hook, or MCP server was loaded.
- The candidate skill is loaded from a temp copy, so the runner can evaluate a proposed SKILL.md without touching the git-tracked one.
- The prompt is the raw user prompt, not `/skill-name`. Whether the skill fires is itself part of what is being measured. Record it from the transcript (a `Skill` tool call naming the skill, or a `Read` of its SKILL.md) as a separate `triggered` field, separate from the quality score.
- Run a baseline arm (no skill loaded) on iteration 0 and keep it. Every later score is reported as delta over baseline. This is what `claude plugin eval --ablation with-without` does natively.

### Stage 2: judge

One `claude -p` per run, in a separate process, with a different tool set (read-only) and ideally a different model from the one under test.

```
claude -p --bare --model <judge-model> \
  --add-dir <run-dir> --tools "Read,Glob,Grep,Bash(python *)" \
  --json-schema "$(cat grade.schema.json)" \
  --output-format json \
  "$(cat judge.md) ..."
```

Fairness rules, all enforced by the runner and not by prose:

- The judge sees the case prompt, the workdir outputs, the transcript, and the assertions. It does **not** see the SKILL.md, the iteration number, or the other candidate's output. It cannot know which version it is grading.
- Grading is assertion-anchored. Each case ships a list of objective checks ("output file exists and parses as CSV", "row count equals 42", "no `TODO` in the deliverable"). Anything checkable with a script is checked by a script in the runner, before the judge runs, and the judge only grades what needs judgment. The `skill-creator` grader agent already articulates the standard: burden of proof on the assertion, no partial credit, cite evidence.
- Structured output via `--json-schema` so the runner never parses free text.
- Judge and judge prompt are frozen per experiment. Changing the judge starts a new experiment; scores across judges are never compared.
- Optional pairwise mode for subjective skills: present two outputs blind, in both orders, and count a win only if it wins both orders. This removes position bias and is more reliable than absolute scores for writing-style skills.

### Stage 3: propose

One `claude -p` with a capable model, whose only job is to emit a new SKILL.md.

- Inputs: current SKILL.md, the train-set grades with evidence, transcripts of the failing train runs, history of prior candidates and their train scores.
- Not inputs: held-out grades, the judge prompt, the assertion text for held-out cases.
- Output: the full new SKILL.md via `--json-schema`, plus a one-paragraph hypothesis ("failures cluster on multi-page inputs; added a pagination step"). The hypothesis goes in the commit message so the git log reads as a lab notebook.
- Constraints in the prompt: generalize, don't enumerate test cases; keep it under a line budget; explain the why rather than adding MUSTs. The `skill-creator` improve prompt is the right starting text.

### The gate

After stage 2, the runner decides deterministically whether to accept the candidate:

```
accept iff  mean(train)   >= best_train   + margin
        and mean(holdout) >= best_holdout - tolerance
        and triggered_rate >= baseline_triggered_rate
```

Accepted candidates are committed. Rejected ones are kept in `runs/` for the editor's history but never become the parent of the next proposal. Stop after K iterations, K consecutive rejections, or a cost ceiling.

### The meta-skill

A SKILL.md in this repo that teaches an interactive Claude Code session to drive the loop. It should be thin: pick or write cases with the user, run `refinery/loop.py`, read the report, discuss. It should **not** run stages inline as subagents, because then the orchestrating session's context leaks into every run and the transcripts are no longer hermetic. The runner is the boundary.

## Build plan

1. **Runner skeleton.** `run.py` and `judge.py` as standalone CLIs against one case, writing the directory layout above. Verify hermeticity by asserting on the `system/init` event (loaded tools, plugins, MCP servers) across two runs.
2. **Assertion format and scripted checks.** YAML with `kind: file_exists | regex | script | judge`. Scripted kinds run in the runner; only `judge` kinds go to stage 2.
3. **Judge calibration.** Before trusting the judge, hand-grade ~20 runs and compare. Report agreement. Re-run until agreement is above an agreed floor; this is the one place a human must be in the loop and it happens once per experiment, not once per iteration.
4. **Loop and gate.** `loop.py` with train/holdout, N runs per case, margin and tolerance from the manifest, git commit per accepted candidate.
5. **Meta-skill.** SKILL.md that explains the workflow and points at the runner.
6. **Reuse, don't rewrite, when it unblocks:** if `claude plugin eval` graduates from early access, stages 1 and 2 can become a thin wrapper around it, keeping stage 3 and the gate here.

## Open questions

- Which model to pin for the run stage vs the judge. Recommendation: run on whatever the users of the skill actually use; judge on the most capable model available, since judge quality caps everything else.
- Whether to support the Claude Agent SDK instead of shelling out. Shelling out is simpler and keeps `claude plugin eval` compatibility; the SDK matters only if the run stage needs in-process hooks.
- Cost. With 3 runs per case, 10 cases, one baseline arm, and 5 iterations, an experiment is on the order of 180 agent runs plus as many judge runs. Budget flags on every process and a per-experiment ceiling in the manifest are non-negotiable.

## How this was researched

Read the full `skill-creator` skill source (SKILL.md, `run_loop.py`, `run_eval.py`, `improve_description.py`, `aggregate_benchmark.py`, the grader, comparator, and analyzer agent prompts, and the JSON schemas). Inspected `claude --help`, `claude plugin --help`, and `claude plugin eval --help` on CLI 2.1.267. Cross-checked flags against the official headless, CLI reference, skills, settings, sandboxing, and Agent SDK docs. Web survey of the frameworks and projects listed above, with the caveats marked inline.
